de.fhdo.lemma.service.open_api/src/de/fhdo/lemma/service/open_api/LemmaDataSubGenerator.xtend

```diff
@@ -0,0 +1,197 @@
+package de.fhdo.lemma.service.open_api
+
+import de.fhdo.lemma.data.Context
+import de.fhdo.lemma.data.DataFactory
+import de.fhdo.lemma.data.DataField
+import de.fhdo.lemma.data.DataStructure
+import de.fhdo.lemma.data.ListType
+import io.swagger.v3.oas.models.OpenAPI
+import io.swagger.v3.oas.models.media.ArraySchema
+import io.swagger.v3.oas.models.media.Schema
+import org.apache.commons.lang3.StringUtils
+import org.slf4j.LoggerFactory
+import de.fhdo.lemma.data.DataModel
+
+/**
+ * This class is responsible for handling the generation of a LEMMA data model from an
+ * OpenAPI file in the JSON or YAML format.
+ *
+ * @author <a href="mailto:jonas.sorgalla@fh-dortmund.de">Jonas Sorgalla</a>
+ */
+class LemmaDataSubGenerator {
+    /**
+     * Map of all created DataStructures during a generation. The key contains the full-qualified
+     * name while the value contains the actual data structure created.
+     */
+    val createdDataStructures = <String, DataStructure>newHashMap
+    /**
+     * Map of all created Lists during a generation. The key contains the full-qualified
+     * name while the value contains the actual list structure created.
+     */
+    val createdListStructures = <String, ListType>newHashMap
+    /** Factory to actually create and manipulate a LEMMA DataModel */
+    val DATA_FACTORY = DataFactory.eINSTANCE
+    /** Predefined instances of the data model, version, and context
+     * which are enriched with the data structures from the OpenAPI source
+     */
+    val myDataModel = DATA_FACTORY.createDataModel
+    val myVersion = DATA_FACTORY.createVersion
+    val myContext = DATA_FACTORY.createContext
+    /** Seperator is used to build the full qualified names during the generation */
+    val seperator = "."
+    /** OpenAPI schema which will be used as source for generation */
+    val OpenAPI openAPI
+    /** SLF4j Logger */
+    val static logger = LoggerFactory.getLogger(LemmaDataSubGenerator)
+    /** Location where the generated file is written */
+    val String targetFolder
+
+    new(OpenAPI openAPI, String genPath, String dataFilename) {
+        super()
+        logger.debug("Creating new Data Sub Generator...")
+        this.openAPI = openAPI
+        this.targetFolder = genPath + dataFilename
+    }
+
+    def DataModel generate() {
+        logger.debug("Starting generation...")
+        logger.debug("Initializing model instance...")
+        initilize()
+        logger.debug("...data model initialized!")
+        logger.debug("Creating data structures...")
+
+        openAPI?.components?.schemas?.forEach[
```

Comment:
I see that `openAPI` may be null because the caller passes `null` to the constructor. If you distrust the caller in this regard, let the constructor immediately throw an Exception when `openAPI === null` and remove the null-safe access operators here entirely.